.chloggen/clr-runtime.yaml

```diff
@@ -0,0 +1,22 @@
+# Use this changelog template to create an entry for release notes.
+#
+# If your change doesn't affect end users you should instead start
+# your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: new_component
+
+# The name of the area of concern in the attributes-registry, (e.g. http, cloud, db)
+component: clr
```

Comment:
Will these metrics be supported on .NET Framework? If so, `dotnet` not accurate? I like `dotnet` though if it is acceptable.

Reply:
I like dotnet. I expect there are lots of places where we use "dotnet" or ".NET" in a broad sense of the entire platform/ecosystem so I wouldn't expect users to be confused. For example you can find both .NET 8 and .NET Framework downloads by following the download link at "https://dotnet.microsoft.com/"

Reply:
Now I'm having second thoughts. I see the java runtime metrics are labeled "jvm" and not "java". I am worried that while "dotnet" might be more recognizable, it may also be ambiguous. Does "dotnet.runtime.xyz" feel too verbose?

Reply:
Can the ambiguity be addressed with a subnamespace (`dotnet.mono`, `dotnet.framework.`) when it's important ?

The litmus test is whether we'd change anything at all in the metric definition.
E.g. if we don't expect CPU utilization to be different, having `dotnet.cpu.` seems enough. 
If we do (I'd be surprised!), then we should have some form of `dotnet.core`, `dotnet.fx` ...

If  metrics are very similar,  but we'd like to introduce some minor differences between metrics and make them more future-proof, we can add a runtime flavor as an  attribute. E.g. `dotnet.runtime.type|flavor|etc`=`core|framework|mono...`

Then we can easily document what the runtime is, what's the difference between metrics, but also can add runtime flavor-specific attributes to metrics.

Reply:
Thinking more about `dotnet.runtime.type|etc` - it could be a good candidate for a resource attribute configured by the OTel SDK by default.

Reply:
> Can the ambiguity be addressed with a subnamespace (dotnet.mono., dotnet.framework.) when it's important ?

Our north star is for .NET to be same everywhere you go. 

Some versions of some variants may be missing some features. These details change over time, and we avoid exposing them in the public surface.

Reply:
> I also feel that the people who consume these low-level metrics will be familiar with the term CLR

I'm a little more skeptical than you are :) I've had many experiences answering questions from folks who were looking at metrics but didn't have a good understanding of what the metrics meant. I think these metrics get exposed to wide audiences just by virtue of being common and available, not necessarily because they are would useful in a specific situation or the observer understood them. I'd guess maybe ~70% of viewers of these metrics will recognize 'CLR' and ~90% would recognize 'dotnet'.

> Our north star is for .NET to be same everywhere you go.

Agreed :)

> E.g. if we don't expect CPU utilization to be different, having dotnet.cpu.

My understanding is that CPU shouldn't be treated as a runtime metric at all, instead we'd want [process.cpu.time](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/process-metrics.md#metric-processcputime)?

But I know that was just an example so maybe if we pick a different example like the threadpool, conceivably some higher layer component could also have a separate threadpool and we'd end up with something like:

dotnet.thread_pool.xyz
aspnet.thread_pool.xyz (this is a purely fictional future thread pool)


I've been continuing to think on it and dotnet.xyz seems like it would fairly clear. Does anyone have objections to "dotnet" as the term we'd move forward with?

Reply:
dotnet seems good to me.

Reply:
I can accept `dotnet` if that's the consensus. On reflection, I don't think the specific runtime implementation matters. If a mono runtime added these metrics, they'd follow the conventions for naming and attributes. They can also choose not to implement the spec. 

Arguably, "dotnet" is a product name rather than a runtime name, though, and these fall under "runtime" metrics in the conventions. That distinction may not matter here, but perhaps someone from the semantic conventions working group needs to weigh in on this as well.

Reply:
> My understanding is that CPU shouldn't be treated as a runtime metric at all, instead we'd want [process.cpu.time](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/process-metrics.md#metric-processcputime)?

@trask @jack-berg do you happen to remember what were the reasons to have [JVM-specific metrics](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/runtime/jvm-metrics.md#metric-jvmcputime) for CPU?

Reply:
check out https://github.com/open-telemetry/opentelemetry-specification/issues/3490 (and maybe https://github.com/open-telemetry/opentelemetry-specification/issues/2291)

and semi-related https://github.com/open-telemetry/opentelemetry-specification/pull/2388#pullrequestreview-896811797 and https://github.com/open-telemetry/opentelemetry-specification/pull/2388#issuecomment-1056064977

may be worth revisiting and documenting more clearly whatever decision is made around this

IIRC one of the influencing factors in JVM metric group decision was that it seemed like the least controversial approach

Reply:
thanks a lot for the context!

If I understand correctly the TL;DR of these discussions is that `process.*` and `system.*` metrics are intended to be reported by an external observer such as collector which does not know runtime specifics, can't add runtime-specific attributes, etc. 

It's actually documented here

https://github.com/open-telemetry/semantic-conventions/blob/a2c9cff852cd63cd9b4e04fab651b147e3cccea1/docs/system/process-metrics.md?plain=1#L9-L16

If .NET did implement `process.cpu.time` it would it some way conflict or duplicate a metric collected on the OS level and prevent .NET from adding any extra information that would not be available on the OS level.

Reply:
Thanks that was different than my previous understanding of how this would done but its np, I want to follow the convention. To double check I've understood this correctly as a concrete example I'm looking at how the runtime would report cpu time on Windows. Based on this I'm expecting the runtime metrics would:

- include a metric "dotnet.cpu.time"
- its values would be populated by a call to the win32 function [GetProcessTimes](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getprocesstimes)
- it would include a single attribute [process.cpu.state](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/process.md) using the 'system' value for kernel CPU time and the 'user' value for user CPU time?

Assuming that is correct I assume similar logic would apply for a metric such as "dotnet.memory.virtual" that calls OS APIs to determine total process committed virtual memory?

Reply:
Yep! that's what I'd like us to try to achieve, see some details below

> dotnet.cpu.time:

- not sure about win32 - why not `Process.GetCurrentProcess().UserProcessorTime` and friends so that's cross-platform?  The goal is to report what runtime thinks about itself, no need to match what OS reports exactly.
- reusing `process.cpu.state` is interesting idea. The .NET terminology would be `user` and `privileged`, right? We could still stick it into `process.cpu.state` for now. We should not use `system` if it does not apply to .NET. Let's start with it, but I'd prefer https://github.com/open-telemetry/semantic-conventions/issues/1139 and will follow up with the semconv community. 
- it should also be useful to report `dotnet.cpu.count` so it'd be possible to calculate normalized utilization.

Reply:
> Assuming that is correct I assume similar logic would apply for a metric such as "dotnet.memory.virtual" that calls OS APIs to determine total process committed virtual memory?

Not sure I understand the OS API part - we should report what runtime knows about itself. If someone needs OS perspective, they can capture things on the OS/container level.

What I'd like to explore is reporting different parts of memory consumption within the same metric name but different dimensions.

Let's see if we can do it - let's start with heap-only memory
`dotnet.heap.used` would contain something like:
- `dotnet.heap.type` = `gc|non-gc`
- `dotnet.heap.generation|space|area|?` = `gen-0|gen-1|gen-2|loh|poh` - should not be set when it does not apply 
- where would the gc-fragmentation go? is it per generation or per all generations? should it be another value of `dotnet.heap.generation = gc-fragmentation`?

Not sure about fragmentation, but such metric looks fine (to me) otherwise.

Let's try to do `dotnet.memory.used` - it would contain something like:
- `dotnet.memory.pool=`native|heap-gc|heap-non-gc|stack|...`
- `dotnet.memory.sub_pool` = `heap-gen-0|heap-gen-1|heap-gen-2|loh|poh|...`
- same concern with gc-fragmentation

Doing this probably does not make sense since we effectively need different attributes for different memory pools.

Reply:
> not sure about win32 - why not Process.GetCurrentProcess().UserProcessorTime

I mentioned the win32 API to mean that the data our counter would provide returns the same value that anyone would observe calling that OS API. The runtime doesn't have some alternative concept about CPU usage that produces different values than what the OS provides. In terms of the actual implementation, yes we'd make it cross-platform and call `Process.GetCurrentProcess().UserProcessorTime`, which itself is a wrapper over different OS APIs depending on platform.

> The .NET terminology would be user and privileged, right?

I think we have to make a choice whether to match the OTel naming convention for process.cpu or match .NET's convention for API terminology. I assumed if we were using the 'process.cpu.state' attribute name then we'd also want to match OTel's convention for how the categories were named. Do you think its better for each runtime to use its own naming scheme? Kernel vs. user time isn't a runtime specific concept.

> it should also be useful to report dotnet.cpu.count so it'd be possible to calculate normalized utilization.

dotnet.cpu.count would report Environment.ProcessorCount?

> dotnet.heap.type = gc|non-gc

I'm not sure if you meant "non-gc" as in:
1. all the other committed virtual memory in the process that isn't the GC heap
2. .NET has a heap which stores managed objects whose name is 'non-gc heap'

If you meant (1) we might want to change the to 'Other' to avoid confusion.
If you meant (2) I have no problem with it in principle but currently there is no BCL API that exposes the information so we probably can't get this implemented in .NET 9.

> dotnet.heap.generation|space|area|

'generation' is probably the right term to pick. Its consistent with other .NET APIs such as `GC.GetGCMemoryInfo().GenerationInfo`. 

Because GC.GetGCMemoryInfo().TotalCommitted is larger than the sum of all the generations I assume again we'd want some 'Other' category that captures all the committed memory attributed to the GC that is not part of a specific generation.

> where would the gc-fragmentation go? is it per generation or per all generations?

If you are looking at `process.runtime.dotnet.gc.heap.fragmentation.size` from the current metrics, that is referring to per-generation fragmentation. Each value there represents the subset of the memory reported with dotnet.heap.generation=X that is not being used to store objects.

> dotnet.memory.used

This feels like this is basically identical to the 'dotnet.heap.used' except with different names? I'm not sure what the value of having both would be?

Reply:
> I think we have to make a choice whether to match the OTel naming convention for process.cpu or match .NET's convention for API terminology. I assumed if we were using the 'process.cpu.state' attribute name then we'd also want to match OTel's convention for how the categories were named. Do you think its better for each runtime to use its own naming scheme? Kernel vs. user time isn't a runtime specific concept.

Enums are open and the `process.cpu.state` is in the process of being redesigned to be used across different metrics - https://github.com/open-telemetry/semantic-conventions/pull/1026 and it's on the way to support extensibility. 

It would be great to use dotnet terminology on dotnet-specific metrics and the `cpu.state|mode` will allow it. 

Until https://github.com/open-telemetry/semantic-conventions/pull/1026 is merged, we can use `process.cpu.state` here.

Reply:
> dotnet.cpu.count would report Environment.ProcessorCount?

Oh, you just opened the new world to me via https://github.com/dotnet/runtime/issues/48094.

To calculate utilization, someone needs to do `(cpu.time[ts2] - cpu.time[ts1])/(ts2 - ts1) / processor_count` but since processor_count is not accurate, it'd be quite confusing if someone configured 0.1 processor quota. 

I.e. my CPU count would show 1, I'd see CPU utilization of 10% which actually means 100% (on my 0.1 quota).

Reply:
> since processor_count is not accurate

Haha - I'd say there isn't a single agreed upon concept that everyone means when they say 'processor count' :) It sounds like it should be straightforward until you get into it.

As a practical matter for implementation within .NET 9 Environment.ProcessorCount is probably the only readily available definition we have. If we think its good we could use it, otherwise we can leave that metric out.

Reply:
> As a practical matter for implementation within .NET 9 Environment.ProcessorCount is probably the only readily available definition we have. If we think its good we could use it, otherwise we can leave that metric out.

Without count it's not possible to normalize at all. E.g.  `(cpu.time[ts2] - cpu.time[ts1])/(ts2 - ts1)` could be 1000% - 10 processors busy 100% of the time - it's probably a bad sign if I have 10 cpus, but just fine if I have 15.

So I have no means to understand what's the relative utilization in my system or create any alerts at all.

It sounds like `Env.ProcessorCount` is still as accurate as .NET can be for whole CPU numbers. So reporting it in a metric would still be better than nothing. And maybe the metric could become more precise in the future versions of .NET (it'd be slightly breaking of course)?

Reply:
I believe this is resolved (`dotnet`,  and cpu count)